utils/resb/src/bundle.rs

```diff
@@ -0,0 +1,326 @@
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! The `bundle` module provides data structures for working directly with the
+//! contents of a resource bundle.
+//!
+//! WARNING: This module is not suitable for use at runtime due to its reliance
+//! on `std` and `alloc` and therefore not intended for general deserialization
+//! of resource bundles. Rather, it is intended to be used in development-time
+//! tools for working with bundles.
+
+extern crate alloc;
+use alloc::{
+    borrow::Cow,
+    collections::{btree_map, BTreeMap},
+};
+
+use crate::MASK_28_BIT;
+
+/// A tree-like collection of data [`Resource`]s primarily intended for storing
+/// locale and other internationalization data for [ICU] (International
+/// Components for Unicode).
+///
+/// [ICU]: https://icu.unicode.org/
+#[derive(Debug)]
+pub struct ResourceBundle<'a> {
+    name: Cow<'a, str>,
+    root: Resource<'a>,
+    is_locale_fallback_enabled: bool,
+}
+
+impl<'a> ResourceBundle<'a> {
+    /// Makes a new resource bundle with the specified resource at its root.
+    pub fn new(name: Cow<'a, str>, root: Resource<'a>, is_locale_fallback_enabled: bool) -> Self {
+        Self {
+            name,
+            root,
+            is_locale_fallback_enabled,
+        }
+    }
+
+    /// Gets the name of the resource bundle.
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    /// Gets the root resource in the resource tree.
+    pub fn root(&self) -> &Resource {
+        &self.root
+    }
+
+    /// Returns `true` if fallback is enabled for this resource bundle.
+    ///
+    /// A resource bundle storing locale data may omit some data in order to
+    /// reduce duplication, allowing fallback to more general locales which
+    /// use the same values.
+    pub fn is_locale_fallback_enabled(&self) -> bool {
+        self.is_locale_fallback_enabled
+    }
+}
+
+/// A data resource within a [`ResourceBundle`].
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum Resource<'a> {
+    /// A well-formed UTF-8 string.
+    String(Cow<'a, str>),
+
+    /// A heterogeneous list of resources, ordered by insertion.
+    Array(Vec<Resource<'a>>),
+
+    /// A set of key-resource pairs, sorted lexically by key.
+    Table(Table<'a>),
+
+    /// A slice of arbitrary binary data.
+    Binary(Cow<'a, [u8]>),
+
+    /// A 28-bit integer.
+    ///
+    /// May be interpreted as either signed or unsigned depending on consumer
+    /// expectations. See [`Int28`] for further details.
+    Integer(Int28),
+
+    /// A list of 32-bit integers, ordered by insertion.
+    IntVector(Vec<u32>),
+}
+
+/// A table of [`Resource`]s indexed by a string [`Key`].
+///
+/// Resources may be accessed either by key or by integer index.
+#[derive(Debug)]
+pub struct Table<'a> {
+    // Table entries are sorted by key, so a BTreeMap is preferred over a
+    // HashMap, which makes no ordering guarantees.
+    map: BTreeMap<Key<'a>, Resource<'a>>,
+}
```

Comment:
what's the value of this new type over `type Table = BTReeMap` ? You seem to be just exposing it's methods.

Reply:
At one point, I had `impl Index` for `Table`, which required newtype, but I removed that since I wasn't sure it was useful and failed to reconsider whether `Table` needed to be a newtype. This can be addressed.

Reply:
Addressed in https://github.com/unicode-org/icu4x/pull/4267.